### PR TITLE
feat(lsp): add support for semantic tokens

### PR DESCRIPTION
**Pull request by smolck**
Relevant part of the spec: https://microsoft.github.io/language-server-protocol/specifications/specification-current/#textDocument_semanticTokens

### CODE REVIEW

**Comment by mjlbach** [comment]
You need to add the capability here where we resolve server capabilities (do a real check, this was just a dirty fix)

```diff
diff --git a/runtime/lua/vim/lsp/protocol.lua b/runtime/lua/vim/lsp/protocol.lua
index 6e98ea1f4..27cbde164 100644
--- a/runtime/lua/vim/lsp/protocol.lua
+++ b/runtime/lua/vim/lsp/protocol.lua
@@ -1016,6 +1016,7 @@ function protocol.resolve_capabilities(server_capabilities)
   general_properties.document_range_formatting = server_capabilities.documentRangeFormattingProvider or false
   general_properties.call_hierarchy = server_capabilities.callHierarchyProvider or false
   general_properties.execute_command = server_capabilities.executeCommandProvider ~= nil
+  general_properties.semantic_tokens_full = true
 
   if server_capabilities.renameProvider == nil then
     general_properties.rename = false
```


Adding this, seems like you're missing textDocument as part of the params you're sending:
```
[ DEBUG ] 2021-03-14T23:57:58-0700 ] /usr/local/share/nvim/runtime/lua/vim/lsp/rpc.lua:391 ]	"rpc.send.payload"	{  id = 2,  jsonrpc = "2.0",  method = "textDocument/semanticTokens/full",  params = {    uri = "file:///Users/michael/Repositories/rust/strings/src/main.rs"  }}
[ DEBUG ] 2021-03-14T23:57:58-0700 ] /usr/local/share/nvim/runtime/lua/vim/lsp/rpc.lua:492 ]	"decoded"	{  error = {    code = -32602,    message = 'Failed to deserialize textDocument/semanticTokens/full: .: missing field `textDocument`; {"uri":"file:///Users/michael/Repositories/rust/strings/src/main.rs"}'  },  id = 2,  jsonrpc = "2.0"}

```

**Comment by Congee** [comment]
Ping

Is there any update for this exciting feature?

**Comment by smolck** [comment]
@Congee I'm happy to continue working on this, but I wasn't quite sure exactly where we wanted to go with it, specifically how to use the tokens by default. My initial idea was to have semantic highlighting, but that feels more like something for an external plugin to deal with (analogous to nvim-treesitter using the builtin tree-sitter things to add highlighting etc.) and isn't as easy to deal with as it would initially seem, and so I'm a little lost at this point.

Maybe a good default functionality would be to store the tokens in a cache and handle creating/cleaning up that for buffers etc.? Then a plugin could retrieve them for use in highlighting or otherwise . . . I'm not sure.

**Comment by Congee** [comment]
@smolck thank you for the update. Yeah maybe the semantic highlighting feature could be solely implemented by a plugin like https://github.com/jackguo380/vim-lsp-cxx-highlight when nvim-treesitter still is not robust for now. Yet after searching through the some neovim communities, this PR is the only relevant work i found that the builtin neovim lsp can make use of.

Thank you for your pioneer work anyway.

**Comment by shadmansaleh** [comment]
@smolck I wanted to do this kind of things entirely from external plugin with #14322 but you and @mjlbach were against it :) . I think treesitter [highlighter](https://github.com/neovim/neovim/blob/master/runtime/lua/vim/treesitter/highlighter.lua) is in core so is regex highlighter . It might not be too bad to do semantic highlighting directly from core .

**Comment by smolck** [comment]
> I think treesitter [highlighter](https://github.com/neovim/neovim/blob/master/runtime/lua/vim/treesitter/highlighter.lua) is in core so is regex highlighter . It might not be too bad to do semantic highlighting directly from core .

Hmm that's a good point; I guess nvim-treesitter just takes care of parser install and such, but the main functionality is in core, didn't think about that. In any case, after a quick discussion w/mjlbach on gitter I think I know where to go with this now, should be able to move things along this evening if all goes well. 

TL;DR is we'll have a way to retrieve the tokens via a function given a client & bufnr, for plugins to do with as they wish. That would enable me to make a plugin for semantic highlighting to flesh that out, and then maybe get it merged into core eventually if we decide we want to have semantic highlighting in core.

**Comment by shadmansaleh** [comment]
> I think I know where to go with this now, should be able to move things along this evening if all goes well.

Cool 👍

> TL;DR is we'll have a way to retrieve the tokens via a function given a client & bufnr, for plugins to do with as they wish. That would enable me to make a plugin for semantic highlighting to flesh that out, and then maybe get it merged into core eventually if we decide we want to have semantic highlighting in core.

Just saw the conversation on gitter . I agree this pr can just add this and later prs can add highlighter if needed .

TBH I don't understand why people are looking for this . TS already provides language aware highlighting 🤷‍♂️

**Comment by Congee** [comment]
@shadmansaleh nvim-treesitter still is not robust enough that I have to frequently do `:e`. Some languages like Haskell or Scalaw does not have a query file for treesitter, either. :cry:

**Comment by shadmansaleh** [comment]
> vim-treesitter still is not robust enough that I have to frequently do :e.

true . But this is also very new .

> Some languages like Haskell or Scalaw does not have a query file for treesitter, either. 😢

I feel like in genaral it might be easier to find a TS parser then a lsp server that supports semantic highlighting . Since building a TS parser takes a lot less effort then a lsp server. there seems to a haskell [parser](https://github.com/tree-sitter/tree-sitter-haskell) but I think it's unmaintained . seeing [this](https://github.com/haskell/lsp/pull/314) I think semantic tokens aren't yet supported by haskell language server . the situation seems same for scala there's an unmaintained TS parser and the lsp server doesn't support semantic highlighting . Though I'm not 100% sure because I don't use these languages :)

**Comment by disrupted** [comment]
> there seems to a haskell [parser](https://github.com/tree-sitter/tree-sitter-haskell) but I think it's unmaintained .

what makes you think it's unmaintained? No offense, I just had a brief look and it looks pretty active to me, last commit is under 1 month ago.

**Comment by smolck** [comment]
> TBH I don't understand why people are looking for this . TS already provides language aware highlighting 🤷‍♂️

I think I have a bit of an understanding of why, but rather than try to explain it myself (which I'm not sure I can do well), I'll just link to [this issue on nvim-treesitter](https://github.com/nvim-treesitter/nvim-treesitter/issues/81) and specifically [this comment on that issue by the author of semshi](https://github.com/nvim-treesitter/nvim-treesitter/issues/81#issuecomment-731710514).

From looking at the end of that issue, it seems tree-sitter does support semantic highlighting, at least in some form. However, I'm sure there are valid usecases for LSP-based semantic highlighting, and perhaps at this point and/or in some cases it is better than tree-sitter highlighting, I don't know.

Regardless, none of this means we shouldn't expose semantic *tokens* to users/plugin authors to do with as they please. The question of whether or not semantic highlighting should be included in core is not something this PR addresses; instead the plan is to only make such a feature possible, whether in a plugin or in core itself, by exposing semantic tokens to users/plugin authors.

**Comment by shadmansaleh** [comment]
> what makes you think it's unmaintained? 

Sorry didn't try to offend anyone 
I was looking at the [list](https://github.com/nvim-treesitter/nvim-treesitter#supported-languages) on nvim-treesitter repo and haskell and scala didn't have check mark and didn't have a maintainer name that's why I thought it was unmaintained .

> , I'll just link to this issue on nvim-treesitter and specifically this comment on that issue by the author of semshi.

Thanks for the links I'll read them :)

**Comment by theHamsta** [comment]
@Congee have you tried the Haskell PR on nvim-treesitter? 

Semantic token will in case be a complement to reged highlighting or Tree-Sitter since the two latter can be used until LSP finished indexing. Maybe I have to see whether there is a way to conceal this feature with Tree-Sitter. It would be cool if Tree-Sitter would know when it needs to do something or whether to use the semantic highlighting when it's available

**Comment by Congee** [comment]
Yeah, There's still additional work needed to have Semshi like highlighting.

@disrupted the Haskell treesitter integration had been unmaintained over 2+ years until [recently](https://github.com/tree-sitter/tree-sitter-haskell/graphs/code-frequency), the README of nvim-treesitter hasn't caught up yet :sweat_smile: 

@theHamsta just tried the latest revision https://github.com/tree-sitter/tree-sitter-haskell/commit/2e33ffa3313830faa325fe25ebc3769896b3a68b, It works for a few lines of code with the only feature incremental selection. And TSModule says highlight and indent are not supported yet.

I am happy to see the community getting active on all these issues :grin:

**Comment by clason** [comment]
Two things (that were mentioned before but only in passing):

1. If tree-sitter based highlighting -- which has been and still is in _very_ active development -- isn't robust yet, what makes you think that semantic highlighting -- which is still at the WIP stage -- is going to be a replacement for production use any time soon! (Not that I don't have full confidence in @smolck, but I sense unreasonable expectations building up again and want to make sure they are managed).
2. Semantic highlighting in general doesn't make sense as a _replacement_ for regex or treesitter highlighting; LSP is challenging to make responsive enough for it, especially on large projects. (People already complained about tree-sitter not being fast enough and showing flicker!) On the other hand, LSP can do things that tree-sitter can't (easily), namely highlight _based on information from other files_ -- for example, if a variable is defined as `const` in `foo.h` and you use it in `bar.c`, it can get a different highlight in the latter to distinguish it from non-`const` variables. So I think it is important to design semantic highlighting support from the ground up to be a supplement to rather than a replacement of whatever "base" highlighting (regex or tree-sitter) is used.

File: runtime/lua/vim/lsp/buf.lua [hunk 1]
```diff
@@ -46,6 +46,11 @@ local function request(method, params, handler)
   return vim.lsp.buf_request(0, method, params, handler)
 end
 
+function M.semantic_tokens_full()
+  local params = util.make_text_document_params()
```
Comment (theHamsta) [inline runtime/lua/vim/lsp/buf.lua:50]: `  local params = { textDocument = util.make_text_document_params() }`

File: runtime/lua/vim/lsp/semantic_tokens.lua [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+    local bit = get_bit(x, i)
+    if bit == 1 then
+      table.insert(modifiers, 1, modifiers_table[i + 1])
+    end
+  end
+
+  return modifiers
+end
+
+function M._handle_semantic_tokens_full(client_id, bufnr, response)
```
Comment (theHamsta) [inline runtime/lua/vim/lsp/semantic_tokens.lua:23]: since the module is already called semantic tokes I guess you can drop it to make the function names more terse

Comment (smolck) [reply runtime/lua/vim/lsp/semantic_tokens.lua:23]: @theHamsta So like `_handle_full()`? Or what did you have in mind exactly?

File: runtime/lua/vim/lsp/semantic_tokens.lua [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+  end
+
+  local client = vim.lsp.get_client_by_id(client_id)
+  return client.request('textDocument/semanticTokens/full', params, handler)
+end
+
+function M.on_refresh()
+  local bufnr = vim.fn.bufnr()
+  -- TODO(smolck): Just do the active clients for the current buffer?
+  -- If so, how to get those?
```
Comment (williamboman-pp) [inline runtime/lua/vim/lsp/semantic_tokens.lua:85]:
```lua
vim.lsp.buf_get_clients({bufnr})
```

Comment (theHamsta) [reply runtime/lua/vim/lsp/semantic_tokens.lua:85]: On refresh is triggered by a particular language server so we should also use this server that requested a refresh: https://github.com/theHamsta/neovim/blob/33de6be20b21a250a9374d25bd0fb9e6cb288664/runtime/lua/vim/lsp/semantic_tokens.lua#L113

File: runtime/lua/vim/lsp/semantic_tokens.lua (hunk 2 above)
Comment (theHamsta) [reply runtime/lua/vim/lsp/semantic_tokens.lua:23]: I would prefer it. Especially for `'lsp.semantic_tokens'.get_semantic_tokens()` a `'lsp.semantic_tokens'.get()` would feel better. But I `get` is feels also a bit unsafe for external users since they must know when to request tokens. Maybe a subscription to token changes would be better?

I also played a bit around with this PR: https://github.com/theHamsta/neovim/blob/81f20e363c95437b12eec600eb003a9b54b5d7d4/runtime/lua/vim/lsp/semantic_tokens.lua#L1

File: runtime/lua/vim/lsp/semantic_tokens.lua [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+      table.insert(tokens[prev_line + 1], #tokens, {
+        start_col = prev_start,
+        length = data[i + 2],
+        type = token_type,
+        modifiers = modifiers
+      })
+    else
+      tokens[prev_line + 1] = {
+        {
+          start_col = prev_start,
```
Comment (theHamsta) [inline runtime/lua/vim/lsp/semantic_tokens.lua:52]: why not use the same names as from the LSP spec?
```
    local token = {
      line = line,
      start_char = start_char,
      length = data[i + 2],
      type = token_type,
      modifiers = modifiers
    }
```

File: runtime/lua/vim/lsp/semantic_tokens.lua [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+  local legend = client.server_capabilities.semanticTokensProvider.legend
+  local token_types = legend.tokenTypes
+  local token_modifiers = legend.tokenModifiers
+  local data = response.data
+
+  local tokens = {}
+  local prev_line, prev_start = nil, 0
+  for i = 1, #data, 5 do
+    local delta_line = data[i]
+    prev_line = prev_line and prev_line + delta_line or delta_line
```
Comment (theHamsta) [inline runtime/lua/vim/lsp/semantic_tokens.lua:34]: Just `line`? The line is not really previous but also current

File: runtime/lua/vim/lsp/semantic_tokens.lua [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+    local delta_start = data[i + 1]
+    prev_start = delta_line == 0 and prev_start + delta_start or delta_start
+
+    -- data[i+3] +1 because Lua tables are 1-indexed
+    local token_type = token_types[data[i + 3] + 1]
+    local modifiers = modifiers_from_number(data[i + 4], token_modifiers)
+
+    if delta_line == 0 and tokens[prev_line + 1] then
+      table.insert(tokens[prev_line + 1], #tokens, {
```
Comment (theHamsta) [inline runtime/lua/vim/lsp/semantic_tokens.lua:43]: if and else branch are essentially the same code. Maybe?
```lua
    local token = {
      line = line,
      start_char = start_char,
      length = data[i + 2],
      type = token_type,
      modifiers = modifiers
    }
    tokens[line + 1] = tokens[line + 1] or {}
    table.insert(tokens[line + 1], token)
```

File: runtime/lua/vim/lsp/semantic_tokens.lua (hunk 4 above)
Comment (clason) [reply runtime/lua/vim/lsp/semantic_tokens.lua:52]: (yes, please -- the closer we stick to the spec, the less confusion)

**Comment by neovim-discourse** [comment]
This pull request has been mentioned on **Neovim Discourse**. There might be relevant details there:

https://neovim.discourse.group/t/lsp-and-semantic-token-support/784/2

**Comment by jackhub** [comment]
any update here?

**Comment by smolck** [comment]
@jackhub Haven’t touched it in a while for a few reasons, one of which being I’ve been waiting on a review from @mjlbach. I do need to get around to @theHamsta’s reviews though, which I should be able to do tomorrow.

**Comment by tom-anders** [comment]
@smolck @theHamsta Do you need some help with testing? I could test it with clangd, if you haven't already.

**Comment by justinmk** [comment]
continued at https://github.com/neovim/neovim/pull/15723
